<!doctype html>
<html>
  <head>
	 <meta name="description" content="Reproducible saturation based peak calling" />
	 <meta name="keywords" content="ChIP-Seq, peak calling, software, bioinformatics, computational biology, genetics, epigenetics" />
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link rel="shortcut icon" href="favicon.ico" />
    <link rel="icon" type="image/png" href="img/favicon-192x192.png" sizes="192x192">
    <link rel="apple-touch-icon" sizes="180x180" href="img/apple-touch-icon-180x180.png">
    <title>ChIP-seq peak caller</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
	 <img src="img/Q_main_header.png" align="top" width='100%'/>
	 <br><br>
	 <p class="header">
    Q is a fast saturation-based ChIP-seq and
    <a href="http://www.ncbi.nlm.nih.gov/pubmed/25751057" target="_idr">ChIP-nexus</a>    
    peak caller.
    Q works well in conjunction with the irreproducible discovery rate
    (<a href="http://statistics.berkeley.edu/sites/default/files/tech-reports/790.pdf" target="_idr">IDR</a>)
    procedure.
    Q was extensively tested on publicly available ChIP-seq datasets from 
    <a href="http://genome.ucsc.edu/ENCODE/" target="_encode">ENCODE</a> and shown to perform well with respect to reproducibility of the called peak set, consistency of the peak sets with respect to 
    predicted transcription factor binding motifs contained in them, as well as overall run time. Q is implemented in C++ making use of the 
    <a href="http://www.seqan.de" target="_seqan">SeqAn</a> library.
	  <br> <br>
	 There are a number of useful features for the primary analysis of ChIP-seq data.	 
	 Q can be run with or without data from a  control experiment.
	 Duplicate reads are removed on the fly without altering the original BAM file, and the number of duplicated reads is then shown in Q's output.
    The average fragment length of the sequencing library,
    which is an essential parameter for peak calling and for downstream analysis,
    is estimated automatically from the data.
    This is done by examing the vector of read start positions along individual chromosomes and
    calcuting the shift that is associated with the smallest Hamming distance. This procedure yields
    an equivalent estimation of the average fragment length as the cross-correlation plot of
    <a href="http://compbio.med.harvard.edu/Supplements/ChIP-seq/" target="_spp">SPP</a>
    but is approximately three times faster.
    As a part of this procedure, Q also calculates the relative strand cross-correlation coefficient
    (<a href="http://genome.ucsc.edu/ENCODE/qualityMetrics.html#chipSeq">RSC</a>),
    which allows a global quality assessment of the enrichment.
    In addition Q offers its own quality metrics,
    which can be used for trouble-shooting and quality control of the results.
    If desired, Q also generates fragment coverage profiles which can be uploaded to
    <a href="http://genome.ucsc.edu/">UCSC's genome browser</a>,
    where they can be displayed in the context of other related data such as 
    for example ChIP-seq data for histone modifications and cofactors or expression data.
    <br> <br>
    ChIP-nexus, an extension of the ChIP-exo protocol, requires less input DNA and
enables selective PCR duplicate removal using random barcodes. We developed 
a comprehensive software package for ChIP-nexus data that exploits the
random barcodes used in the ChIP-nexus protocol. Furthermore, we developed
bespoke methods to estimate the length of the protected region resulting from
protein-DNA binding as well as for peak calling.
    <br> <br>
    If you use Q for ChIP-seq, please cite:
	<ul>
	  Hansen et al. (2015) Saturation analysis of ChIP-seq data for reproducible identification of binding peaks.
 Genome Research, 25:1391-400.
	  [PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/26163319">26163319</a>]
	</ul>
    </p>
    <hr/>
    <div class="wrapper">
      <header>
        <ul>
          <li><a href="https://github.com/charite/Q/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/charite/Q/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="http://github.com/charite/Q">Fork On <strong>GitHub</strong></a></li>
        </ul>
      <p class="view"><a href="http://github.com/charite/Q">View the Project on GitHub</a></p>
      <p class="view"><a href="installation_nexus.html">Installation</a></p>
      <p class="view"><a href="tutorial.html">Tutorial for ChIP-seq</a></p>      
      <p class="view"><a href="tutorial_chip_nexus.html">Tutorial for ChIP-nexus</a></p>      
      </header>
      <section>
        <h1>NEWS</h1>
        <dl>
          <dt>02/08/2016: New release v1.2.0 is available</dt>
          <dd>Additional software is provided for the preprocessing of ChIP-nexus data.
	  When Q is executed in a newly added nexus-mode the width of the protected region will be estimated
	  and used for peak calling with parameters optimized for ChIP-nexus data.
	  </dd>
        </dl>     
        <dl>
          <dt>13/11/2015: Bug fix for issue 6 in Q v1.1.1</dt>
          <dd>The function writeBedGraph caused a corrupt memory error. This bug is fixed in Q version 1.1.1.</dd>
        </dl>        
        <dl>
          <dt>18/09/2015: New release v1.1.0 is available</dt>
          <dd>Q is now able to read paired-end data. In addition linking problems with zlib during istallation have been addressed and a pre-compiled binary of Q has been added to the repository.</dd>
        </dl>
        <dl>
          <dt>12/01/2015: First release of Q</dt>
          <dd>The first release of Q is now available on GitHub.</dd>
        </dl>
        <br>
        <br>
        <br>                
      </section>
      <footer>
        <p>This project is maintained by <a href="https://charite.github.io/team_hansen.html">Peter Hansen</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
  </body>
</html>
