<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link rel="shortcut icon" href="favicon.ico" />
    <link rel="icon" type="image/png" href="img/favicon-192x192.png" sizes="192x192">
    <link rel="apple-touch-icon" sizes="180x180" href="img/apple-touch-icon-180x180.png">
    <title>ChIP-seq peak caller - Installation</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  
         <body>
         <div class="round-button">
         <a href="index.html">
           <img src="img/Home_Icon.svg" alt="Home" />
         </a>
         </div>
         <br><br>
         <div class="wrapper">

         <h1>Q-nexus: Installation of preprocessing tools</h1>
         
			<p>
			Q-nexus comprises two applications designed to preprocess ChIP-nexus FASTQ and alignment (BAM) files prior to
			peak calling with the Q application (using a special flag for ChIP-exo and ChIP-nexus data). This installation
			guide covers the installation of the preprocessing applications, flexbar and nexbar. 
			This installation guide is subdivided into the following sections:
			<ul>
			 <li><a href="#h2_installation_for_linux" >Installation for Linux</a></li>
			  <ul>
			   <li><a href="#h3_precompiled_binary" >Precompiled binary</a></li>
			   <li><a href="#h3_installation_from_source" >Compile Q from source</a></li>
				<li><a href="#h3_compile_preprocessing_tools" >Compile preprocessing tools from source</a></li>
			  </ul>
			 <li><a href="#h2_install_ucscs_converter_tools" >Install UCSC's converter tools</a></li>
			</ul>
			</p>         
      
         <hr/>

         <h2 id="h2_installation_for_linux">Installation for Linux</h2>
         
         <h3 id="h3_precompiled_binary" >Precompiled binary</h3>         
              
         <p>
         We provide precompiled binaries for Linux x86_64. The main Q application as well
         as flexcat and nexcat can be downloaded as a part of the 
         <a href="https://github.com/charite/Q">Q github repository</a>. Clone this repository in order
         to download all the files (cd to the directory into which you would like to download Q and enter the
          following command): 
         <pre><code>git clone https://github.com/charite/Q.git</code></pre>
         
		You will now find executable (binary) files for the applications Q, flexcat, and nexcat as well as
		a python script that can be used to carry out the entire process of preprocessing, alignment, and PCR duplicate
		removal at once. Users can use an aligner of their choice. In order to use the script (which uses bowtie and samtools), 
		it is necessary to install Bowtie, Samtools 
         and Python. 
         If you are using Ubuntu, this can be done using the following command
			</p>
         <pre><code>sudo apt-get install samtools bowtie python</code></pre>

               
	
                  
         <p>         
         Open your hidden personal configuration file using a text editor of your choice
         (for example <code>gedit</code> or <code>kate</code>).
			</p>

         <pre><code>gedit ~/.profile</code></pre> 
         
         <p>
         Add  the following line to your profile:
			</p>
			
			<pre><code>export PATH=$PATH:$HOME/Q/bin</code></pre>
			
         <p>
			For the first time you have to reload your profile using:
         </p>
			 
			<pre><code>source ~/.profile</code></pre>
			         
         <p>
         Try the following commands (which will display help messages) to test the installation.         
         </p>
			
			<pre><code>Q --help</code></pre>
			<pre><code>flexcat --help</code></pre>
			<pre><code>nexcat --help</code></pre>
			<pre><code>Q/scripts/preprocess.py --help</code></pre>
			
         <h3 id="h3_installation_from_source">Compile Q from source</h3> 

         <h4>Preliminaries</h4>

         <p>         
         Q is written in C++ and makes use of several external libraries.
         Q uses
         <a href="http://www.seqan.de/">SeqAn</a>
         for BAM file I/O,
         and the SeqAn library is provided together with the Q source code and does not need to be installed separated.
         The zlib, bz2, and gomp libraries are needed for operating with the (compressed) BAM files and for multiprocessor support.
         Here, the dev versions of the libraries should be installed
         (which are guaranteed to have the header files needed for compilation).
         In addition, the boost library is used by Q for certain statistical calculations.
         </p>

         <p>
         On Ubuntu/Debian install all required packages using the following command line:
         </p>
         
         <pre><code>sudo apt-get install zlib1g-dev libbz2-dev libboost-dev libgomp1</code></pre>  

         
         <p>
         On other linux distrubtions, install the libraries using the command appropriate to your system.
         We have tested Q using the following compiler and library versions.
			<ul>
			<li>g++: 5.1</li>
			<li>libboost-dev: 1.49.0.1</li>
			<li>zlib1g-dev: 1:1.2.7.dfsg-13</li>
			<li>libbz2-dev: 1.0.6-4</li>
			</ul>
			</p>
			
			<h4>Installation</h4>
			
			<p>
			Change into the cloned Q directory
			and use the following command to check if your system has the required libraries and to create a Makefile:
			</p>
			
			<pre><code>./configure</code></pre>

			<p>
			Assuming no error messages are produce, Q can now be compiled. There are two ways of doing so.  
			You can  install the executable in a system wide directory (if you have write permissions, e.g., because you are root):
			</p>

			<pre><code>make; make install</code></pre>			

			<p>
			Alternatively, to install Q in another directory, use the following to create the Makefile:
			</p>
			
			<pre><code>./configure --prefix=/path/to/my/directory</code></pre>
			
			<p>
			Entering <code>make;make install</code> will now install Q into <code>/path/to/my/directory/bin/Q</code>.
			</p>

			<p>
			The above will compile a version of Q will all of the library code included as well as debugging symbols. To create
			a release target, you can use the following command:
			</p>
			
			<pre><code>make install-strip</code></pre>

			<p>
			In order to make the compile binary available on the command line
			you will have to adjust your $PATH variable
			(see section for the <a href="#h3_precompiled_binary" >precompiled binary</a>).
			</p>
			
		
			<h3 id="h3_compile_preprocessing_tools">Compile preprocessing tools from source</h2>
			<p>
			The following section describes how to compile the preprocessing tools from source. It has been tested on Ubuntu 14.04.
			</p>
			
				<p>
			Install G++ 5.1 or newer
			</p>
			<pre><code>sudo apt-get install g++-5</code></pre>

				<p>
			Install Cmake 3.2 or newer
			</p>
			<pre><code>sudo apt-get install software-properties-common
sudo add-apt-repository ppa:george-edison55/cmake-3.x
sudo apt-get update
sudo apt-get install cmake</code></pre>
				<p>
			Install Boost
			</p>
			<pre><code>sudo apt-get install libboost-all-dev</code></pre>

			
			<p>
			Clone the source code from the github repository if not done already:
			</p>

			<pre><code>git clone https://github.com/charite/Q.git</code></pre>


			<p>
			Change in to the Q source directory:
			</p>
			
			<pre><code>cd Q/src
git clone https://github.com/seqan/seqan.git seqan2
cd seqan2
git checkout seqan-v2.0.1</code></pre>

			<p>
			Change into the preprocessing directory and execute the build script
			</p>
			<pre><code>cd ../preprocessing
./build_linux_release.sh</code></pre>
		 

		 
			
			<h2 id="h2_install_ucscs_converter_tools">Install UCSC's converter tools</h2>

         <p>	
         Q generates BedGraph files for the fragment coverage profiles on the fly. This section describes the installation of
         several UCSC executables that will be needed (See the <a href="tutorial.html#visualization">tutorial</a> for details).
         Sometimes the BigWig format is more suitable and we will implement the generation of BigWig files in near future.
         By then two tools from the UCSC command line utilities are needed for the conversion from BedGraph to BigWig format.
         For downloading change to the bin directory within your home directory and enter the following command lines.
         </p>

         <pre><code>wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/fetchChromSizes</code></pre>
         
         <pre><code>wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedGraphToBigWig</code></pre>

         <p>	
         Now you have to make the utilities executable using the following command:
         </p>
         
         <pre><code>chmod u+x fetchChromSizes bedGraphToBigWig</code></pre>

         <p>		
         If you have added <em>/home/bin</em> to your $PATH variable,
         the utilities should now be available.
         </p>
			
			
			
			
			
			
         				
			<br>
			<br>
			<br>						
			<br>
			<br>
			<br>						
			<br>
			<br>
			<br>						


         </div>
    <script src="javascripts/scale.fix.js"></script>
  </body>
</html>
