<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link rel="shortcut icon" href="favicon.ico" />
    <link rel="icon" type="image/png" href="img/favicon-192x192.png" sizes="192x192">
    <link rel="apple-touch-icon" sizes="180x180" href="img/apple-touch-icon-180x180.png">
    <title>Q-nexus Manual</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

  </head>
  
         <body>
         <div class="round-button">
         <a href="index.html">
           <img src="img/Home_Icon.svg" alt="Home" />
         </a>
         </div>
         <img src="img/Nexus-Tools_tutorial_header.png" align="top" width='100%'/>
         <br><br><br>
         <div class="wrapper">
         <h1>Q-nexus Manual</h1>
         
         <p>
        Q-nexus is a peak caller designed for the analysis of ChIP-nexus data.
        <a href="http://www.ncbi.nlm.nih.gov/pubmed/25751057" target="_new1">ChIP-nexus</a>
		(chromatin immunoprecipitation experiments with nucleotide resolution through exonuclease, 
			unique barcode and single ligation), is an extension of ChIP-seq and ChIP-exo that utilizes an efficient 
			DNA self-circularization step during library preparation. This allows protein-DNA binding sites
			to be characterized to base-pair resolution. Q-nexus additionally can be used to analyze ChIP-exo data. This
			tutorial explains how to install and use Q-nexus, and presents a complete example using the
			a dataset in which the transcription factor dorsal was investigated in Drosophila melanogaster. 
			</p>
			<p>
			The entire Q-nexus pipeline requires three programs.
			<ol>
<li><strong>flexcat</strong> performs adapter clipping and identifies the random barcodes added to sequence fragments by the Q-nexus protocol (the barcodes arewritten to the ID line of the filtered FASTQ file that is produced by flexcat). The resulting FASTQ file is aligned to the target genome using an external program such as bowtie.</li>
<li>The <strong>nexcat</strong> program then removes PCR duplicates from the resulting BAM file but retains distinct reads that have been mapped to  identical 
 positions on the basis of the random barcodes (this is one of the major advantanges of the ChIP-nexus protocol). Finally,</li>
<li><strong>Q-nexus</strong> performs peak calling using an extension of the 
			<a href="http://www.ncbi.nlm.nih.gov/pubmed/26163319" target="q_new">Q algorithm</a> for ChIP-seq peak calling.			</li>
</ol>
			</p>

			<p>
			This Manual is subdivided into the following sections:
			<ul>
			 <li><a href="#preprocessing" >Preprocessing</a></li>
			<ul>
			 <li><a href="#adapter_trimming_and_fixed_barcode_detection" >Adapter trimming and fixed barcode detection</a></li>
			 <li><a href="#mapping" >Mapping the reads to the reference genome</a></li>
			 <li><a href="#post_mapping_filtering" >Post-mapping filtering</a></li>
			 <li><a href="#executing_the_script" >Executing the script</a></li>
			 <li><a href="#visualization_and_statistics" >Visualization and Statistics</a></li>
			</ul>
			 <li><a href="#peak_calling" >Running Q in nexus-mode</a></li>
			<ul>
				 <li><a href="#binding_characteristics" >Binding characteristics</a></li>
				 <li><a href="#peak_calling" >Peak calling</a></li>
			</ul>
			</ul>
			</p>
						
			<hr/>
			
		   <h2 id="preprocessing">Preprocessing</h2>
			
			
		   <h3 id="adapter_trimming_and_fixed_barcode_detection">flexcat: Adapter Trimming and filtering for fixed Barcode</h3>
			<p>
			In the first step, we process the raw FASTQ file generated by the ChIP-nexus experiment in order
			to perform adapter clipping and simultaneously identify the Q-nexus random barcodes. 
			We note that the random barcode is followed in the 3' direction by
			a four nucleotide fixed barcode (See Figure 1A of the Q-nexus manuscript). Reads that do not have this
			fixed barcode are likely to be artefacts and are removed by flexcat. The fixed barcode and the random barcode
			 are removed from	the sequence but the random barcode is added to the ID line of the output FASTQ file.
</p>			
<p>ChIP-exo data do not have a fixed or random barcode. Preprocessing therefore involves only adaptor trimming. </p>
<p> As an optional step for both ChIP-nexus and ChIP-exo data, quality trimming can be performed.</p>			
			 The following options are used for this application
                </p>
	
        <table>
          <tbody>
              <tr>
            <th>Option</th><th>Description</th>
          </tr>
              <tr>
            <th>-tl 5</th><th>Trim 5 bases from the 5" end of each read, this removes the random barcode from the read sequence</th>
          </tr>
              <tr>
            <th>-tt</th><th>Store trimmed sequences in the corresponding id, this is neccessary to process the trimmed random barcode at a later stage</th>
          </tr>
              <tr>
            <th>-t</th><th>Tag sequences with removed adapters</th>
          </tr>
              <tr>
            <th>-ml 0</th><th>Set the final minimum length of reads that will be accepted, default is 0</th>
          </tr>
              <tr>
            <th>-er 0.2</th><th>Error rate for adapter matching, default is 0.2</th>
          </tr>
              <tr>
            <th>-ol 4</th><th>Minimum required overlap for adapter matching, default is 4</th>
          </tr>
              <tr>
            <th>-app</th><th>Allow 1 base error for fixed barcode matching</th>
          </tr>
              <tr>
            <th>-tnum</th><th>Number of threads, set according to tnum script parameter</th>
          </tr>
              <tr>
            <th>-ss</th><th>Display processing speed in reads per second</th>
          </tr>
              <tr>
            <th>-fr [N]</th><th>Process only first N reads, default is all reads</th>
          </tr>
              </tbody>
        </table>
            <p>flexcat requires two input files that specify the adapter sequences to be trimmed as well as the fixed barcode.            
            The adapter sequences can be specified in the file <strong>adapters.fa</strong>. For example,
            the following file can be used for the adaptor in the original Q-nexus protocol ( which is located in the 
            Q/data/ directory. This file could for example look like this</p>
             <pre><code>
>adapter1
AGATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC
>adapter2
GATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC
>adapter3
ATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC
>adapter4
TCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCC
</code></pre>
			<p>For single-end reads, adapter*.1 is omitted. The adapters in this example file are chosen to match 
                ChiP-nexus protocol. If you use another another protocol, for example ChiP-Exo, this file 
                 needs to be modified accordingly.
			</p>
                 <p>ChiP-nexus reads have in contrast to ChiP-Exo reads two barcodes at the 5"-end of each read. The first barcode is a random barcode of length 5, the second barcode immediately follows 
        the first barcode and is a fixed barcode of length 4. The fixed barcode is used to identify valid ChiP-nexus reads. Therefore Reads at this processing step will be filtered for the fixed barcode,
        and only those who match the fixed barcode in Q/data/barcodes.fa with an allowed matching error of 1 base will be used for further processing. For the standard ChiP-nexus protocol, the 
        barcodes.fa file looks like this
		    </p>
                          <pre><code>>matched_barcode
CTGA             </pre></code>

			<h3 id="mapping" >Mapping the reads to the reference genome</h3>
			<p>Mapping is done using <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie 1.1.2</a>. Although there is already Bowtie 2 available, the authors 
                of Bowtie say, that Bowtie 1.1.2 is better suitable for short reads. The following options are used for bowtie
			</p>
			        <table>
          <tbody>
              <tr>
            <th>Option</th><th>Description</th>
          </tr>
              <tr>
            <th>-k 1</th><th>report up to 1 valid alignment per read</th>
          </tr>
              <tr>
            <th>-m 1</th><th>only map reads that are uniquely mappable</th>
          </tr>
              <tr>
            <th>--chunkmbs 512</th><th>each thread can use 512 MBs memory</th>
          </tr>
              <tr>
            <th>--strata --best</th><th>See <a href="http://bowtie-bio.sourceforge.net/manual.shtml#bowtie-options-best description">Bowtie manual</a> for detailed </th>
          </tr>
              <tr>
            <th>-S</th><th>Output file format is SAM</th>
          </tr>
              <tr>
            <th>-p</th><th>Number of threads, set according to tnum script parameter</th>
          </tr>
              <tr>
            <th>-v 2</th><th>Set verbosity level to 2, however this output will only be displayed, if the 
 script is executed with --verbose
                         </th>
          </tr>
              </tbody>
        </table>

		   

		   <h3 id="post_mapping_filtering">Post-mapping filtering</h3>
	   <p>The mapped reads need to be further processed, because for ChiP-nexus we want to remove reads that are mapped to the same position and have the same random barcode. This 
           enabled us to distinguish PCR artifact duplicates from real duplicates. For this preprocessing stage, the binary nexus-pre is used with the following settings
	   </p>

			        <table>
          <tbody>
              <tr>
            <th>Option</th><th>Description</th>
          </tr>
              <tr>
            <th>-fc [REGEX]</th><th>regular expression to remove chromosomes from the calcuation of the qfragment-length-distribution</th>
          </tr>


              </tbody>
        </table>



         <h3 id="executing_the_script">Executing the whole process</h3>

		   <p>
            We provide a script called preprocess.py, that enabled you to run all the above mentioned preprocessing steps 
               with one command. We recommend the following directory structure.</p>
			        <table>
          <tbody>
              <tr>
            <th>Directory</th><th>Description</th>
          </tr>
              <tr>
            <th>data</th><th>a directory where all the input and output data of the preprocessing resides</th>
          </tr>
              <tr>
            <th>data/fastq</th><th>raw reads</th>
          </tr>
              <tr>
            <th>data/genomes</th><th>reference genomes</th>
          </tr>
              <tr>
            <th>data/preprocessed</th><th>directory where all the output data of the preprocessing goes</th>
          </tr>

              </tbody>
        </table>
        <p>
             Now you can start the preprocessing pipeline be executing this command</p>
                                       <pre><code>~/Q/scripts/preprocess.py ~/data/fastq/input_reads_ID01.fastq ~/data/genome/ref_genome.fa --clean --output_dir ~/data/preprocessed/ID01 --verbose --overwrite
CTGA             </pre></code>
        <p>The following table explains the options that can be used with this script</p>
			        <table>
          <tbody>
              <tr>
            <th>Option</th><th>Description</th>
          </tr>
              <tr>
            <th>[INPUT FILE]</th><th>path the the input file with raw reads (fastq or fastq.gz)</th>
          </tr>
              <tr>
            <th>[GENOME FILE]</th><th>path to the reference genome file (fasta or fasta.gz)</th>
          </tr>
              <tr>
            <th>--exo</th><th>don't do barcode trimming (use this option if you process ChiP-Exo reads)</th>
          </tr>
              <tr>
            <th>--clean</th><th>delete intermediate files, such as the mapping output generated by bowtie</th>
          </tr>
              <tr>
            <th>--overwrite</th><th>overwrite existing output files, if this option is not used, preprocessing steps will be skipped if the corresponding output file already exists</th>
          </tr>
              <tr>
            <th>--verbose</th><th>display output of the different preprocessing stages on screen</th>
          </tr>
              <tr>
            <th>--bowtie_location [PATH]</th><th>specify the path to the bowtie executable, this is neccessary if bowtie is not in the path and for all windows installations</th>
          </tr>
              <tr>
            <th>--num_threads [N]</th><th>use N threads, default is 4</th>            
          </tr>
              <tr>
            <th>--output_dir [PATH]</th><th>directory where the output files will be created, default is the current working directory</th>
          </tr>
              <tr>
            <th>--filter_chromosomes [REGEX]</th><th>regular expression to remove chromosomes from the calcuation of the qfragment-length-distribution</th>
          </tr>

              </tbody>
        </table>

		   <h3 id="visualization_and_statistics">Visualization and Statistics</h3>
			<p>
			The preprocessing pipeline generates a file which is called [INPUT_PREFIX]_duplication_rate.txt. This file containts information 
                about how many reads with a specific coverage rate exist. We also supply and R-script that can generate the following graph out of 
                this txt file.
			</p>
		   
		   <center>	
		   <img src="img/duplication_rate.png" width="100%"/>
			</center>
			<p>
            On the x-axis of this graph is the coverage rate, the left y-axis show the number of reads. The right y-axis show the percentage of 
                non unique duplet reads out of all duplet reads. The discrimination of unique and non unique duplicates is for the first time possible, because 
                of the random barcodes used by the ChiP-nexus protocol. The preprocessing output can be further analyzed using the tool 
                <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>.
			</p>
			<p>
			The preprocessing pipeline also generates a file with statistical information about the preprocessed data. It could look like this
                          <pre><code>Total reads	17102794
Filtered reads (-fc Option)	0
Mapped reads	9616348
Non mappable reads	2082353
Non uniquely mappable reads	5404093
After barcode filtering	6891129	 (-2725219)
PCR duplication rate	0.283394
Total duplet reads	1274581
Estimated fragment Length	18</pre></code>
			</p>
	
 		   <br>
 		   <br>		   			 
			 
		   <h2 id="peak_calling">Running Q in nexus-mode</h2>
			
		   <h3 id="binding_characteristics">Binding characteristics</h3>
		   <h3 id="peak_calling">Peak calling</h3>

			 
    </div>
    <script src="javascripts/scale.fix.js"></script>
  </body>
</html>