<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link rel="shortcut icon" href="img/Q-logo.png" />
    <title>Tutorial</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  
         <body>
         <div class="round-button">
         <a href="index.html">
           <img src="img/Home_Icon.svg" alt="Home" />
         </a>
         </div>
          <div class="round-button" style="bottom:120px">
         <a href="#top">
           <img src="img/Top_Icon.svg" alt="Home" />
         </a>
         </div>
         <img src="img/Q_tutorial_header.png" align="top" width='100%'/>
         <br><br><br>
         <div class="wrapper">
         <h1>Tutorial</h1>
        
			<p>This tutorial will take users through the process of peak calling with Q and will demonstrate a number of typical steps in the downstream analysis.
			If you have not yet downloaded Q, please see the 
			section on <a href="installation.html">installation</a> before going further.
			</p>
			<p>
			The tutorial is subdivided into the following sections:
			<br>&nbsp;&nbsp;&nbsp;
			<a href="#getting_mapped_chip_seq_reads" >Getting mapped ChIP-seq reads</a>
			<br>&nbsp;&nbsp;&nbsp;
			<a href="#running_q" >Running Q</a>
			<br>&nbsp;&nbsp;&nbsp;
			<a href="#output_of_q" >Output of Q</a>
			<br>&nbsp;&nbsp;&nbsp;
			<a href="#fragment_length_estimation" >Fragment length estimation</a>
			<br>&nbsp;&nbsp;&nbsp;
			<a href="#statistics_and_quality_metrics" >Statistics and quality metrics</a>
			<br>&nbsp;&nbsp;&nbsp;
			<a href="#visualization" >Visualization</a>	
			<!--<br>&nbsp;&nbsp;&nbsp;
			<a href="#reproducibility_anlaysis_using_idr">Reproducibility anlaysis using IDR</a>-->	
			</p>
			
			<hr/>
			
		   <h2 id="getting_mapped_chip_seq_reads">Getting mapped ChIP-seq reads</h3>
			<p>
			Q requires as input a file representing the aligned ChIP-seq read data in BAM or SAM format. If the ChIP-seq experiment was performed with a generic IgG control (as is to be recommended),
			there will be one BAM/SAM file for the experiment with the specific antibody ("treatment") and one BAM/SAM file for the experiment with generic IgG ("control").
			If you are starting with unmapped FASTQ files, alignment can be performed with a number
			of programs including <a href="http://bio-bwa.sourceforge.net/">bwa</a> and <a href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie</a>. In general, it is recommended that
			reads that map to multiple positions in the genome (i.e., repetitive sequences) should be removed prior to ChIP-seq analysis (For instance, non-uniquely mapped reads are tagged with 'XT:A:U' by bwa
			for single end data, and can be removed by bowtie using the options '--best -m 1').
			For the purposes of this tutorial, we will use aligned ChIP-seq reads from the 
			<a href="http://genome.ucsc.edu/ENCODE/">ENCODE project</a>,
			it is also possible to download the corresponding BAM files via web browser from
			<a href="http://genome.ucsc.edu/cgi-bin/hgFileSearch?hgsid=396934473_lsCkVpuNAo1IgpRQpFe0KmdprNt5&db=hg19&hgt_tsDelRow=&hgt_tsAddRow=&tsName=&tsDescr=&tsGroup=Any&fsFileType=bam&hgt_mdbVar1=cell&hgt_mdbVal1=HCT-116&hgt_mdbVar2=antibody&hgt_mdbVal2=Pol2-4H8&hgfs_Search=search">
			UCSC</a>.
			Alternatively, the same BAM files can be downloaded using
			<a href="http://www.gnu.org/software/wget/">wget</a>.
			</p>
			
			<pre><code>wget -N ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibTfbs/wgEncodeHaibTfbsHct116Pol24h8V0416101AlnRep*;</code></pre>

			<p>
			According to the
			<a href="http://www.ncbi.nlm.nih.gov/pubmed/22955991" >
			ENCODE guidelines for ChIP-seq</a>, experiments should be accompanied by control experiments. At the
			<a href="http://genome.ucsc.edu/cgi-bin/hgFileSearch?hgsid=396934473_lsCkVpuNAo1IgpRQpFe0KmdprNt5&db=hg19&hgt_tsDelRow=&hgt_tsAddRow=&tsName=&tsDescr=&tsGroup=Any&fsFileType=bam&hgt_mdbVar1=cell&hgt_mdbVal1=HCT-116&hgt_mdbVar2=antibody&hgt_mdbVal2=Pol2-4H8&hgfs_Search=search">
			UCSC ENCODE browser</a>, you can find the contol dataset that corresponds to a given experiment by looking for the <em>controlId</em> in the <em>Additional Details</em> section. 
			For instance, for one experiment we find <strong>controlId=SL3457;</strong> in the Additional Details section. In order to find the corresponding control experiment,
			search for
			<a href="http://genome.ucsc.edu/cgi-bin/hgFileSearch?hgsid=396934473_lsCkVpuNAo1IgpRQpFe0KmdprNt5&db=hg19&hgt_tsDelRow=&hgt_tsAddRow=&tsName=&tsDescr=&tsGroup=Any&fsFileType=bam&hgt_mdbVar1=cell&hgt_mdbVal1=HCT-116&hgt_mdbVar2=labExpId&hgt_mdbVal2=SL3457&hgfs_Search=search" target="_SL3457">
			'Lab provided indentifier' is among 'SL3457'</a>. Again, the BAM files may be downloaded via web browser or wget.
			</p>

			<pre><code>wget -N ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibTfbs/wgEncodeHaibTfbsHct116RxlchV0416101AlnRep1.*;</code></pre>			

			<p>
			In this case, this will leave you with six files, two BAM files for the replicates and one for the control,
			as well as the three corresponding BAM index files.
			If desired, the mapped ChIP-seq data can now be directly visualized using the
			<a href="http://www.broadinstitute.org/igv/" target="_igv">
			Integrative Genome Viewer</a>.
			</p>
			
			<h2 id="running_q" >Running Q</h2>
			
			<p>
			Q can be applied directly to the downloaded datasets.
			Due to Q's statistical model there is no need for downsampling of control data.
			It is generally recommended to remove duplicates (i.e. multiple reads mapping to the same position) before peak calling.
			Duplicate reads can originate from PCR errors during library preparation,
			Q removes duplicates on the fly (without altering the original BAM file), thus obviating the need for using a separate program such as 
			<a href="http://www.htslib.org/">samtools</a> to remove duplicate reads. If you have already removed duplicate reads, no additional adjustments are required for Q. 
			Note that Q allows multithreading.
			In order to use n threads to perform the analysis, pass the argument '--thread-num n' to the command line (e.g., '--thread-num 8' for 8 threads).
			If you want to print additional progress information to the screen use the option '--verbose'.
			The option '--help' will show information about all command line arguments.
		   To perform the basic analysis, you need to indicate the path to the treatment dataset (with the '-t' flag) and the name of the prefix 
			you want for the output files of Q (with the '--out-prefix' flag). If a control experiment was performed, the path to the corresponding BAM file is indicated with the '-c' flag. 
			</p>
			<p>
			In the following example, we perform peak calling
			on replicate 1 and 2 for a ChIP-seq experiment involving RNA polymerase II in HCT-116 cells
			(<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM803474">GSM803474</a>).
			The file names of the corresponding BAM
			files are those from the UCSC ENCODE ChIP-seq browser. The command assumes that
			the treatment and control BAM files are located in the current working directory. 
		   </p>
		   
		   <pre><code>Q -t wgEncodeHaibTfbsHct116Pol24h8V0416101AlnRep1.bam -c wgEncodeHaibTfbsHct116RxlchV0416101AlnRep1.bam -o Hct116Pol24h8Rep1 -wbt -wbc</code></pre>
		
		   <pre><code>Q -t wgEncodeHaibTfbsHct116Pol24h8V0416101AlnRep2.bam -c wgEncodeHaibTfbsHct116RxlchV0416101AlnRep1.bam -o Hct116Pol24h8Rep2 -wbt</code></pre>
		   
  			<p>
			The additional flags '-wbt' and '-wtc' cause the generation of bedGraph files for treatment and control (see below).
			</p>

	      <h2 id="output_of_q">Output of Q</h2>

		   <p>
			After the analysis is finished,  you will find the following files for replicate 1 (and analogous files for replicate 2).
         <table>
          <tbody><tr>
            <th>Filename</th><th>Description</th>
          </tr>
          <tr>
            <td><a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-runinfo.txt">Hct116Pol24h8Rep1-Q-runinfo.txt</a></td>
            <td>Contains all chosen parameters and general information about the run.</td>
          </tr>
          <tr>
            <td><a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-binding-characteristics.txt">Hct116Pol24h8Rep1-Q-binding-characteristics.R</a></td>
            <td>R script that generates a plot for the fragment-length estimation procedure.</td>
          </tr>
          <tr>
            <td><a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-quality-statistics.tab">Hct116Pol24h8Rep1-Q-quality-statistics.tab</a></td>
            <td>Contains useful quality statistics regarding the performance of the corresponding ChIP-seq experiment.</td>
          </tr>
          <tr>
            <td><a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-narrowPeak.tab">Hct116Pol24h8Rep1-Q-narrowPeak.bed</a></td>
            <td>
               The actual result and contains the predicted binding sites.
               Find a detailed description of the narrowPeak format
               <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format12">here</a>.
            </td>
          </tr>
          <tr>
            <td><a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-treatment_head.bedgraph">Hct116Pol24h8Rep1-Q-treatment.bedgraph</a></td>
            <td>
            Contains signal density profile for treatment data which can be loaded as custom track within genome browser of UCSC.
            Find a detailed description of the bedGraph format
            <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1.8">here</a>.
            </td>
          </tr>
          <tr>
            <td><a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-control_head.bedgraph">Hct116Pol24h8Rep1-Q-control.bedgraph</a></td>
            <td>
            Contains signal density profile for the control data.
            </td>
          </tr>          
         </tbody></table>
		   </p>
		   
		   <h2 id="fragment_length_estimation">Fragment length estimation</h2>
		   <p>
		   Q outputs an R script that allows users to visualize the results of the fragment-length estimation procedure (similar to the 
		   "binding characteristics analysis" of
		   <a href="http://compbio.med.harvard.edu/Supplements/ChIP-seq/tutorial.html">SPP</a>). This file can be used to generate a graphic in <a href="http://www.r-project.org/" target="_r">R</a>.
		   In our example, the file
		   <a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-binding-characteristics.txt">Hct116Pol24h8Rep1-Q-binding-characteristics.R</a>
		   is used  to generate the the corresponding plot
		   <a href="img/Hct116Pol24h8Rep1-Q-binding-characteristics.pdf">Hct116Pol24h8Rep1-Q-binding-characteristics.pdf</a>
		   by entering the following command in the shell:
		   
		   <pre><code>Rscript Hct116Pol24h8Rep1-Q-binding-characteristics.R</code></pre>

   		<p>
		   The pdf file for replicate 1 is shown below on the left.
		   If SPP is applied to the same dataset (following duplicate removal by samtools),
		   an analogous, but vertically flipped curve with nearly identical characteristics is generated, shown on the right.
		   </p>
		   
			<center>	
			<iframe src="img/Hct116Pol24h8Rep1-Q-binding-characteristics.pdf#view=fitH" width="350"  height="350">
			</iframe>
			<iframe src="img/Hct116Pol24h8Rep1-SPP-binding-characteristics.pdf#view=fitH" width="350"  height="350">
			</iframe>
			<p>Binding characteristics plot for Hct116Pol2Rep1 for Q (left) and SPP (right)</p>
		   </center>	   

		   <p>
		   The main difference is that Q calculates the <a href="http://en.wikipedia.org/wiki/Hamming_distance" target="_hamming">Hamming distance</a>,
		   whereas SPP calculates the <a href="http://en.wikipedia.org/wiki/Pearson_product-moment_correlation_coefficient" target="_pearson">Pearson correlation coefficient</a>.
		   The shift with the minimum Hamming distance is taken as the estimate of the average fragment length by Q, while the shift with the maximum correlation is taken as the
		   estimate of the average fragment length by SPP.
			The RSC values reflect quality of enrichment and are used as a
		   <a href="http://genome.ucsc.edu/ENCODE/qualityMetrics.html#chipSeq">quality metric for ENCODE data</a>,
		   are almost identical for Q and SPP.
   		The RSC value is a useful metric for trouble shooting and optimization ChIP-seq protocols.
		   We compared the results of the
		   <a href="img/all_bc_plots.pdf">binding characteristics analysis on 38 datasets</a>,
		   and found comparable estimates of fragment length and RSC for Q and SPP.
		   </p>
	   
         <h2 id="statistics_and_quality_metrics">Statistics and quality metrics</h2>

		   <p>
		   Q provides you with feedback about the performance of your ChIP-seq experiment.
		   This is useful for trouble-shooting and enables you to
		   evaluate the quality of your data relative to many published datasets from
		   <a href="http://genome.ucsc.edu/cgi-bin/hgFileSearch">ENCODE</a>.
		   The number of uniquely mapped non redundant reads or hits (<b>UMNRH</b>) that actually map to real binding sites of the probed factor
		   is a good measure for the overall performance of your ChIP-seq experiment, the more the better.
		   This number depends roughly on three factors,
		   the number of uniquely mapped reads, the complexity of your sequencing library and the efficiency of the enrichment.
		   Q provides you with quality metrics for all of these factors.
		   </p>		
		   <p>
		   The complexity of the library is reflected by the <b>duplication rate</b>,
		   which is given as the proportion of uniquely mapped reads that are redundant,
		   i.e. there are at least two reads having identical sequences.
		   Redundant reads can originate from PCR errors during library preparation.
		   A high duplication rate means a low complexity of the library.
		   </p>
		   <p>
		   The efficiency of enrichment is measured as the Q's enrichment score (<b>QES</b>).
		   This is a number between 0 and 1,
		   which reflects the efficiency of the IP step and can be interpreted like a signal-to-noise-ratio.
		   A QES of 1 means that all UMNRH occur at binding positions of the probed factor,
		   whereas a QES of 0 means the opposite.
		   The QES is determined by first determining the observed fraction of all UMNRHs in qfrags
		   and then subtracting the expected fraction,
		   where a UMNRH on the forward strand forms a qfrag with a UMNRH on the reverse strand
		   if the distance is at least one estimated fragment length minus 50
		   and at most one estimated fragment length plus 50.
		   </p>
		   <p>
		   The duplication rate and the degree of enrichment defines
		   what proportion of the sequenced reads read will appear at binding positions.
		   These two factors are combined in an score, called Q's sample score (<b>QSS</b>).
		   It is defined as 1 minus the duplication rate multiplied with the QES. 
		   As the name implies, this is a quality metric for the sample.
		   By multiplying the number of uniquely mapped reads with the QSS
		   the absolute number of hits in signals (<b>ENSH</b>) is estimated.
		   </p>		

		   <p>
		   The tab separated file
		   <a href="downloads/tutorial_q_results/Hct116Pol24h8Rep1-Q-quality-statistics.tab">Hct116Pol24h8Rep1-Q-quality-statistics.tab</a>		
		   contains global information for each individual chromosome and for the whole genome.
		   Most of the column names are self-explanatory and a detailed description is as follows.
		   </p>
		   
		   <table>
		   <tr>
 		    <th align="center"><b>Column</b></th>
 		    <th>Description</th>
		   </tr>
		   <tr>
 		    <td align="center">1</td>
 		    <td>Chromosome name</td>
 		   </tr>
 		   <tr>
 		    <td align="center">2</td>
  		    <td><b><font color="blue">Total number of mapped treatment reads that were used by Q</b></td> 
 		   </tr> 		 
 		   <tr>
 		    <td align="center">3</td>
 		    <td>Number of duplicates that were removed from treatment data</td>
		   </tr>
 		   <tr>
 		    <td align="center">4</td>
		    <td><b><font color="blue">Duplication rate for treatment data</font></b></td>
 		   </tr>
 		   <tr>
 		    <td align="center">5</td>
  		    <td><b><font color="blue">Total number of mapped control reads that were used by Q</b></td> 
 		   </tr> 		 
 		   <tr>
 		    <td align="center">6</td>
  		    <td>Number of duplicates that were removed from control data</td>
 		   </tr>
		   <tr>
 		    <td align="center">7</td>
  		    <td><b><font color="blue">Duplication rate for control data</font></b></td>
 		   </tr>
  		   <tr>
 		    <td align="center">8</td>
 		    <td>Expected number of hits in qfrags</td>
 		   </tr> 
 		   <tr>
 		    <td align="center">9</td>
  		    <td>Observed number of hits in qfrags</td>
  		   </tr>
  		   <tr>
 		    <td align="center">10</td>
   		 <td><b><font color="blue">Q's enrichment score (QES)</font></b></td>
  		   </tr>
  		   <tr>
 		    <td align="center">11</td>
   		 <td><b><font color="blue">Q's sample score (QSS)</font></b></td>
  		   </tr>
  		   <tr>
 		    <td align="center">12</td>
   		 <td><b><font color="blue">Estimated number of hits in signals (ENSH)</font></b></td>
  		   </tr>
  		   <tr>
 		    <td align="center">13</td>
   		 <td>Estimated fragment length</font></td>
  		   </tr>    		  	  		   		             
  		   <tr>
 		    <td align="center">14</td>
   		 <td><b><font color="blue">Relative Strand Cross-correlation coefficient (RSC)</font></b></td>
  		  </tr>
		  </table>
				
		   <p>	
		   The lines for the individual chromosomes are given for completeness
		   and the most important line is the last one in which the results for the whole genome are summarized.
		   In the first column this line contains the prefix for output given via the option '--out-prefix'
		   and not the chromosome name.
		   This is useful when collecting the results from multiple ChIP-seq experiment in
		   <a href="downloads/quality_statistics/quality-satistics-summary.html"> one table</a>.
		   In order to provide you with reference values for the quality metrics introduced above,
		   we analysed 392 datasets from ENCODE.
		   The results are summarized in the table below.
		   </p>
		
		   <table>
		   <tr>
 		    <th></th>
 		    <th>very low</th>
 		    <th>moderate low</th> 		   
 		    <th>moderate high</th>
 		    <th>very high</th> 		   
 		   </tr>			
		   <tr>
 		    <td><b>UMNRH</b></td>
 		    <td><= 11,520,000</td>
 		    <td>11,520,000 <</td> 		   
 		    <td>16,380,000 <</td>
 		    <td>22,770,000 <</td> 		   
 		   </tr>
		   <tr>
 		    <td><b>DR</b></td>
 		    <td><= 0.07</td>
 		    <td>0.07 <</td> 		   
 		    <td>0.13 <</td>
 		    <td>0.23 <</td> 		   
 		   </tr>
		   <tr>
 		   <td><b>QES</b></td>
 		    <td><= 0.13</td>
 		    <td>0.13 <</td> 		   
 		    <td>0.20 <</td>
 		    <td>0.28 <</td> 		   
 		   </tr>
		   <tr>
 		    <td><b>QSS</b></td>
 		    <td><= 0.12</td>
 		    <td>0.12 <</td> 		   
 		    <td>0.16 <</td>
 		    <td>0.21 <</td> 		   
 		   </tr>
		   <tr>
 		    <td><b>ENSH</b></td>
 		    <td><= 2,267,000</td>
 		    <td>2,267,000 <</td> 		   
 		    <td>3,041,000 <</td>
 		    <td>4,326,000 <</td> 		   
 		   </tr>
 		   <caption align="bottom">We determined the following values for
 		   <a href="downloads/quality_statistics/quality-satistics-summary.html ">392 datasets from ENCODE</a>		
 		   :
 		   Number of uniquely mapped no-redundant hits (<b>UMNRH</b>),
 		   Duplication rate (<b>DR</b>),
 		   Q's enrichment score (<b>QES</b>),
 		   Q's sample score (<b>QSS</b>)
 		   and the estimated number of hits in signals (<b>ENSH</b>).
 		   We used the first three quantiles to categorize the metrics into the classes
 		   very low, moderate low, moderate high and very high. See here
 		   <a href="downloads/quality_statistics/pdf">here</a>
 		   for the corresponding binding characteristics plots.
 		   </caption>		  		 		 	 		 	
		   </table>
		   
		   <h2 id="visualization">Visualization</h2>
		   
		   <p>
		   Use Q with the option '-wbt' and '-wbc' to generate
		   <a href="http://genome.ucsc.edu/goldenpath/help/bedgraph.html">BedGraph</a>
		   files for the fragment coverage profile on the fly.
		   The fragment coverage profile is a popular and intuitive representation of
		   ChIP-seq data.
		   BedGraph (and also 
		   <a href="http://genome.ucsc.edu/goldenpath/help/bigWig.html">BigWig</a>)	
		   formated files can be loaded as custom track in UCSC's genome browser,
		   which allows you to present your own data in the context of other related data,
		   for example ChIP-seq data histone modifications or cofactors or expression data.
		   Once you have composed a collection of tracks and found an interesting region
		   you can easily export vector graphics in ps or pdf format,
		   which are especially suited for publication.
		   See the following figure for an illustrative example.
		   </p>

			<center>	
			<iframe src="img/hgt_genome_euro_5689_a8ed60.pdf#view=fitH" width="750"  height="310">
			</iframe>
			<p>
			Example for a figure generated in the UCSC genome browser.
			The upper two tracks display the bigWig files generated in the course of this tutorial.
			In both replicates analyzed, RNA Pol2 binds to the Notch2 promoter in HTC116 cells.
			The publicly available ChIP-seq tracks for chromatin marks and RNA-seq in HCT116 cells show
			that the promoter is active (H3K4me3 (lower green track) and H3K27ac (magenta)) and 
			that the Notch2 gene is transcribed (black track).
			Also, the dark red track, a ChIP-seq for the transcription factor SRF,
			shows that SRF binds to the first intron and the corresponding binding site shows H3K4me1
			(upper green track) and H3K27ac (magenta) marks typical for active enhancers.
			</p>
			</center>

		   <p>
		   For large datasets the BigWig format is more suitable,
		   because it allows smoother browsing.
		   It is possible to deposit BigWig files on a web accessible server and
		   provide UCSC's genome browser with a link to the data.
		   If you do not have your own web-server, it is also possible to upload bigWig files
		   to an online file storage provider, such as Dropbox*.
		   The next steps will guide you through the generation of BigWig files,
		   the upload to Dropbox,
		   the integration as custom tracks within UCSC's genome browser,
		   as well as the customization and the export of figures.
		   </p>
		   
		   <p>
		   We will implement the generation of BigWig files for Q as soon as possible.
		   By then the BedGraph files have to be converted using two command line utilities of UCSC,
		   <em>fetchChromSizes</em> and <em>bedGraphToBigWig</em>.
		   These tools can be 
		   <a href="http://hgdownload.cse.ucsc.edu/admin/exe/">downloaded</a>
		   for Linux and Mac.
		   Please see section on <a href="installation.html">installation</a> before going further.
		   </p>

		   <p>
		   The converter <em>bedGraphToBigWig</em> needs information about the chromosome sizes.
		   Use <em>fetchChromSizes</em> to fetch the proper sizes from UCSC.
		   </p>		   	   				
         <pre><code>fetchChromSizes hg19 > hg19.chrom.sizes</code></pre>

   		<p>
		   Now convert the BedGraph files into BigWig format
		   </p>
         <pre><code>bedGraphToBigWig Hct116Pol24h8Rep1-Q-treatment.bedgraph hg19.chrom.sizes Hct116Pol24h8Rep1-Q-treatment.bigwig</code></pre>		   
	      <pre><code>bedGraphToBigWig Hct116Pol24h8Rep2-Q-treatment.bedgraph hg19.chrom.sizes Hct116Pol24h8Rep2-Q-treatment.bigwig</code></pre>
	      <pre><code>bedGraphToBigWig Hct116Pol24h8Rep1-Q-control.bedgraph hg19.chrom.sizes Hct116Pol24h8Rep1-Q-control.bigwig</code></pre>

         <br>
		   <p>
		   Now upload the files to your Dropbox and create public links. It is important to place the files within the public folder,
		   otherwise a public link cannot be created.
		   </p>		

		   <center>	
		   <img src="img/dropbox_link.png" width="800" heigth="200"/>
		   </center>

         <br>
		   <p>
		   Prepare track definition lines for custom tracks in UCSC's genome browser. Use the public links as <em>bigDataUrl</em>.
		   If you generated your own public links edit the text box below correspondingly.
		   </p>		

<pre><code>track type=bigWig name="Hct116Pol24h8Rep1" description="Fragment Coverage" visibility=full
 bigDataUrl=https://dl.dropboxusercontent.com/u/35942072/BigWig/Hct116Pol24h8Rep1-Q-treatment.bigwig

track type=bigWig name="Hct116Pol24h8Rep2" description="Fragment Coverage" visibility=full
 bigDataUrl=https://dl.dropboxusercontent.com/u/35942072/BigWig/Hct116Pol24h8Rep2-Q-treatment.bigwig
 
track type=bigWig name="Hct116Pol24h8Control" description="Fragment Coverage" visibility=full
 bigDataUrl=https://dl.dropboxusercontent.com/u/35942072/BigWig/Hct116Pol24h8Rep1-Q-control.bigwig 
</code></pre>
         <br>
		   <p>
		   Now go to 
		   <a href="http://genome-euro.ucsc.edu/cgi-bin/hgGateway?redirect=auto&source=genome.ucsc.edu">UCSC's genome browser</a>.
		   After selecting the proper assembly (in our case hg19) select <em>manage custom tracks</em>.
		   On the next page select <em>add custom tracks</em>.
		   Now paste the track definition lines from the text box above into the <em>Paste URLs or data</em> box
		   and select the <em>Submit</em> button. 
		   </p>
		   
   		<center>	
		   <img src="img/ucsc_add_custom_tracks.png" width="800" heigth="200"/>
		   </center>
		   
		   <br>
   		<p>
		   After a short while you will be back on the <em>manage custom tracks</em> page.
		   Then select <em>go to genome browser</em> where your custom tracks should now be displayed.
		   You can customize each track by right-clicking and selecting configure.
		   Add tracks for related data by selecting <em>track search</em> below the tracks.
		   Export the current browser graphic in pdf or eps format by selesting <em>View</em> at the top and then <em>PDF/PS</em>. 
		   </p>
 		   <br>
 		   <br>		   
    </div>
    <script src="javascripts/scale.fix.js"></script>
  </body>
</html>