<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link rel="shortcut icon" href="favicon.ico" />
    <link rel="icon" type="image/png" href="img/favicon-192x192.png" sizes="192x192">
    <link rel="apple-touch-icon" sizes="180x180" href="img/apple-touch-icon-180x180.png">
    <title>Nexus-Tools Preprocessing - Tutorial</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

  </head>
  
         <body>
         <div class="round-button">
         <a href="index.html">
           <img src="img/Home_Icon.svg" alt="Home" />
         </a>
         </div>
         <img src="img/Nexus-Tools_tutorial_header.png" align="top" width='100%'/>
         <br><br><br>
         <div class="wrapper">
         <h1>Q-nexus Tutorial</h1>
         
			<p>
			This tutorial is subdivided into the following sections:
			<ul>
			 <li><a href="#preprocessing" >Preprocessing</a></li>
			<ul>
			 <li><a href="#getting_reads" >Getting Reads</a></li>
			 <li><a href="#adapter_trimming_and_fixed_barcode_detection" >Adapter trimming and fixed barcode detection</a></li>
			 <li><a href="#mapping" >Mapping the reads to the reference genome</a></li>
			 <li><a href="#post_mapping_filtering" >Post-mapping filtering</a></li>
			 <li><a href="#executing_the_script" >Executing the script</a></li>
			 <li><a href="#visualization_and_statistics" >Visualization and Statistics</a></li>
			</ul>
			 <li><a href="#peak_calling" >Running Q in nexus-mode</a></li>
			<ul>
				 <li><a href="#binding_characteristics" >Binding characteristics</a></li>
				 <li><a href="#peak_calling" >Peak calling</a></li>
			</ul>
			</ul>
			</p>
						
			<hr/>
			
		   <h2 id="preprocessing">Preprocessing</h2>
			
		   <h3 id="getting_reads">Getting Reads</h3>
		   
		   
		   
	To start the tutorial a ChIP-nexus dataset that was generated using an antibody against the Drosophila melanogaster
	transcription factor
	Dorsal (See <a href="http://www.ncbi.nlm.nih.gov/pubmed/25751057">He et al., 2015, ChIP-nexus enables improved detection of in vivo transcription factor binding footprints</a>).
<pre><code>wget ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fastq/SRA142/SRA142150/SRX475440/SRR1175698.fastq.bz2
</pre></code>

The Drosophila melanogaster genome (version 3, dm3) can be downloaded from the UCSC Genome Browser site and unzipped. We note 
that for Q-nexus, the reference genome should be contained in a single FASTA file. Since the UCSC Drosophila genome is
sp√ºlit into files for each chromosome, we use the cat command to combine the data into a single FASTA file called dm3.fa.

<pre><code>wget http://hgdownload.soe.ucsc.edu/goldenPath/dm3/bigZips/chromFa.tar.gz
tar xvfz chromFa.tar.gz
cat *.fa > dm3.fa</code></pre>

<p>We recommend the following directory structure.</p>
			        <table>
          <tbody>
              <tr>
            <th>Directory</th><th>Description</th>
          </tr>
              <tr>
            <th>data</th><th>a directory where all the input and output data of the preprocessing resides</th>
          </tr>
              <tr>
            <th>data/fastq</th><th>raw reads</th>
          </tr>
              <tr>
            <th>data/genomes</th><th>reference genomes</th>
          </tr>
              <tr>
            <th>data/preprocessed</th><th>directory where all the output data of the preprocessing goes</th>
          </tr>

              </tbody>
        </table>
<p>Place the above files into these directories (or directly download into them).</p>

			
		   <h3 id="adapter_trimming_and_fixed_barcode_detection">flexcat: Adapter Trimming and filtering for fixed Barcode</h3>
			<p>
	
	
	Execute the following commands (If necessary, alter the path so that the shell finds the flexcat executable):
	 <pre><code>cd data/preprocessed
mkdir SRR1175698
cd SRR1175698
flexcat ../../fastq/SRR1175698.fastq.bz2 -o SRR1175698_out.fastq -a ~/Q/data/adapters.fa -times 5 -t -tt -ss -st -app -tnum 4 -er 0.2 -ol 4 -fm 13
</pre></code>
<p>
flexcat removes the random and fixed ChIP-nexus barcodes, adds information about the random barcodes to toe FASTQ description
line, and performs 3' adapter trimming. The reads are now ready to be aligned to the reference genome. If your mapper supports compressed input files, 
you can also choose your output file to have the ending fastq.gz (Bowtie 1.1.2 does not support compressed input files).
</p>


<h3 id="mapping" >Mapping the reads to the reference genome</h3>
<p>Mapping is done using <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie 1.1.2</a>, which is suitable for
the short reads characcteristic of ChIP-nexus data. Note that bowtie requires an index of the reference
genome, which can be generate using the following command.
			</p>
<pre><code>bowtie-build -o 1 ../../genomes/dm3.fa ../../genomes/dm3</pre></code>
<p>The following options are used for the alignment.</p>
	 <pre><code>bowtie -S -p 4 --chunkmbs 512 -k 1 -m 1 -v 2 --strata --best ../../genomes/dm3 SRR1175698_out.fastq SRR1175698_out.sam</pre></code>
	
			

		   

		   <h3 id="post_mapping_filtering">Post-mapping filtering</h3>
	   <p>The mapped reads need to be further processed, because for ChiP-nexus we want to remove reads that are mapped to the same position and have the same random barcode. This 
           enabled us to distinguish PCR artifact duplicates from real duplicates. 
	   </p>
	run <pre><code>nexcat -fc '(.*)[H|U|M]+(.*)' SRR1175698_out.sam</pre></code>
	
<p>The Drosophila genome build 3 (dm3) contains two "chromosomes" that are used to place sequences that could not be localised to
one of the six canonical chromosomes. ChrU (U="unknown"), chrUextra. There are also "Het" versions of the
chromosomes (e.g., chrXhet) that contains sequences that could not be localized within the chromosome. These chromosomes are usually
removed before downstream analysis for ChIP-seq and ChIP-nexus experiments. We remove the sequences using next cat with the <code>-fc</code>
flag (filter chromosome). A regular expression can be used (in our example, the regular expression removes all chromosomes whose name
includes the upper case letters U or H (from het), M (for mitochondrial)).
 For human data it may be useful to include an underscore ("_") 
 in the regular expression to remove chromosomes such as chr21_gl000210_random, but users should carefully consider this depending
 on the organism and the experimental goals.
 See also <a href="http://ucscbrowser.genenetwork.org/cgi-bin/hgGateway?hgsid=732&clade=insect&org=0&db=0">UCSC explanations</a>.
</p>	
<p>nexcat produces a SAM file from which PCR duplicates have been removed. The file now nbeeds to be sorted and converted
to a BAM file prior to downstream analysis with Q-nexus.</p>	
	
	
	   <pre><code>samtools view -Sb SRR1175698_out.sam
samtools sort SRR1175698_out.bam SRR1175698_out_sorted.bam
samtools index SRR1175698_out_sorted.bam</pre></code>
		
	

         <h3 id="executing_the_script">Executing the whole process using a python script</h3>

		   <p>
            We provide a script called <strong>preprocess.py</strong>, that enabled you to run all the above mentioned preprocessing steps 
               with one command. 
        <p>
             Instead of performing all the steps above separately, you can start the preprocessing pipeline be executing this command</p>
	   <pre><code>~/Q/scripts/preprocess.py ~/data/fastq/input_reads_ID01.fastq ~/data/genome/ref_genome.fa --clean --output_dir ~/data/preprocessed/SRR1175698 --verbose --overwrite</pre></code>

		   <h2 id="visualization_and_statistics">Visualization and Statistics</h2>
			<p>
			The preprocessing pipeline generates a file which is called [INPUT_PREFIX]_duplication_rate.txt. This file containts information 
                about how many reads with a specific coverage rate exist. We also supply and R-script that can generate the following graph out of 
                this txt file.
			</p>
		   
		   <center>	
		   <img src="img/duplication_rate2.png" width="100%"/>
			</center>
			<p>
            On the x-axis of this graph is the coverage rate, the left y-axis shows the number of reads. 
            The right y-axis shows the percentage of 
                non unique duplet reads out of all duplet reads. The discrimination of unique and non unique duplicates is for the first time possible, because 
                of the random barcodes used by the ChiP-nexus protocol. The preprocessing output can be further analyzed using the tool 
                <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>.
			</p>
			<p>
			The preprocessing pipeline also generates a file with statistical information about the preprocessed data. It could look like this
                          <pre><code>Total reads	17102794
Filtered reads (-fc Option)	0
Mapped reads	9616348
Non mappable reads	2082353
Non uniquely mappable reads	5404093
After barcode filtering	6891129	 (-2725219)
PCR duplication rate	0.283394
Total duplet reads	1274581
Estimated fragment Length	18</pre></code>
			</p>
	
 		   <br>
 		   <br>		   			 
			 

			
 <h2 id="peak_calling">Running Q in nexus-mode</h2>
			
			<p>
			The original implementation of Q has been designed for ChIP-seq data.
			Due to the different nature of ChIP-seq and ChIP-nexus data,
			we made some adjustments of Q to ChIP-nexus data.
			Using the switch --nexus-mode will cause Q to use optimized parameter settings.
			</p>
			
        <code><pre>Q --nexus-mode -t ~/data/preprocessed/SRR1175698/SRR1175698_filtered_sorted.bam -o SRR1175698 -s 100 -v</pre></code>

			<p>
			Running Q in nexus-mode will cause Q to keep duplicates (assuming only real and no PCR artifacts for ChIP-nexus data).
			The parameter -l will be determined using a methodology especially designed for ChIP-nexus data.
			Furthermore, the evaluation of predicted binding sites differs from the original implementation of Q.
			Find more detailed informations regarding binding characteristics and peak calling below. 
			</p>
					
		   <h3 id="binding_characteristics">Binding characteristics</h3>

			<p>
			If Q is invoked in nexus-mode and no value for fragment length (-l) is specified,
			 it will be set automatically to the predominant qfrag-length observed in the data.
			 This is done by determining the distribution of qfrag-lengths in the given data
			 as well as in the pseudo-control (see publication).
			 Q will produce an R script containing these distributions along with code for generating the corresponding plot (see below),
			 which can be compiled to a pdf using the following command. 
			</p>

        <code><pre>Rscript SRR1175698-Q-qfrag-binding-characteristics.R</pre></code>

			</p>
			
		 	<p>  
		   <center>	
		   <img src="img/SRR1175698-Q-qfrag-binding-characteristics.jpg" width="100%"/>
			</center>
			The plot above shows on the left hand side the distribution of qfrag-lengths in the original data (black) and in the pseudo-control (grey).
			On the right hand site the difference between these two distributions is shown.
			Q will use the length at which the difference reaches its maximum for -l. 
			</p>

			
		   <h3 id="peak_calling">Peak calling</h3>

			<p>
			...	
			</p>
	
		   <br>
 		   <br>			 
    </div>
    <script src="javascripts/scale.fix.js"></script>
  </body>
</html>